src/mimeparser.rs

```diff
@@ -0,0 +1,1143 @@
+    LocationOnly = 9,
+}
+
+impl Default for SystemMessage {
+    fn default() -> Self {
+        SystemMessage::Unknown
+    }
+}
+
+const MIME_AC_SETUP_FILE: &str = "application/autocrypt-setup";
+
+impl<'a> MimeParser<'a> {
+    pub fn from_bytes(context: &'a Context, body: &[u8]) -> Result<Self> {
+        let mail = mailparse::parse_mail(body)?;
+        let mdns_enabled = context.get_config_bool(Config::MdnsEnabled);
+
+        let mut parser = MimeParser {
+            parts: Vec::new(),
+            header: Default::default(),
+            subject: None,
+            is_send_by_messenger: false,
+            decrypting_failed: false,
+            encrypted: false,
+            signatures: Default::default(),
+            gossipped_addr: Default::default(),
+            is_forwarded: false,
+            context,
+            reports: Vec::new(),
+            is_system_message: SystemMessage::Unknown,
+            location_kml: None,
+            message_kml: None,
+            mdns_enabled,
+            parsed_protected_headers: false,
+        };
+
+        let message_time = mail
+            .headers
+            .get_first_value("Date")?
+            .and_then(|v| mailparse::dateparse(&v).ok())
+            .unwrap_or_default();
+
+        parser.hash_header(&mail.headers);
+
+        // Memory location for a possible decrypted message.
+        let mail_raw;
+
+        let mail = match e2ee::try_decrypt(parser.context, &mail, message_time) {
+            Ok((raw, signatures)) => {
+                // Valid autocrypt message, encrypted
+                parser.encrypted = raw.is_some();
+                parser.signatures = signatures;
+
+                if let Some(raw) = raw {
+                    mail_raw = raw;
+                    info!(context, "decrypted: {:?}", std::str::from_utf8(&mail_raw));
+
+                    let decrypted_mail = mailparse::parse_mail(&mail_raw)?;
+
+                    // Decrypted the mail
+                    decrypted_mail
+                } else {
+                    // Message was not encrypted
+                    mail
+                }
+            }
+            Err(err) => {
+                // continue with the current, still encrypted, mime tree.
+                // unencrypted parts will be replaced by an error message
+                // that is added as "the message" to the chat then.
+                //
+                // if we just return here, the header is missing
+                // and the caller cannot display the message
+                // and try to assign the message to a chat
+                warn!(parser.context, "decryption failed: {}", err);
+                mail
+            }
+        };
+
+        parser.parse_mime_recursive(&mail)?;
+
+        // Handle gossip headers
+        let gossip_headers = mail.headers.get_all_values("Autocrypt-Gossip")?;
+        parser.gossipped_addr =
+            update_gossip_peerstates(context, message_time, &mail, gossip_headers)?;
```

Comment:
this seems wrong.  We can only use and process gossip headers if they are in the encrypted part.